### ISSUE TITLE
Suggestion: Noninferential type parameter usage

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#14829]
We keep getting bugs like this and I keep not finding the original so I'm making a new one so we can find it.

Keywords: ignore type parameter optional inference

### Problem

Often with generics, there will be some locations where a type parameter *should* be inferrable from usage, and *other* places where the type parameter should only be used to *enforce* typechecking. This comes up in a variety of contexts

```ts
class Animal { move }
class Dog extends Animal { woof }

function doSomething<T>(value: T, getDefault: () => T) { }
// Wanted an error here - getDefault() ought to return same type as 'value'
doSomething(new Dog(), () => new Animal());
```

```ts
declare function assertEqual<T>(actual: T, expected: T): boolean;
const g = { x: 3, y: 2 };
assertEqual(g, { x: 3 }); // Forgot y, wanted error
```

### Proposal Sketch

We should be able to mark type parameter consumption sites as being "not eligible for inference". For example, let's say we had a special global type that the compiler knew not to unwrap during inference:
```ts
type NoInfer<T> = T;
```
Then we can annotate usage sites
```ts
function doSomething<T>(value: T, getDefault: () => NoInfer<T>) { }
// Wanted an error here - getDefault() ought to return same type as 'value'
doSomething(new Dog(), () => new Animal());
```
```ts
declare function assertEqual<T>(actual: T, expected: NoInfer<T>): boolean;
const g = { x: 3, y: 2 };
assertEqual(g, { x: 3 }); // Error
```

**Comment by mhegazy** [comment]
Is using multiple type parameters not an option?

```ts
function doSomething<T, U extends T>(value: T, getDefault: () => U) { }

declare function assertEqual<T, U extends T>(actual: T, expected: U): boolean;
```

**Comment by johnfn** [comment]
@mhegazy, speaking from my experience, I never ever would have guessed that 

`function doSomething<T>(value: T, getDefault: () => T) ;`

could be amended to do what I want by changing it to

`function doSomething<T, U extends T>(value: T, getDefault: () => U) ;`

Multiple type parameters are an option! It's just that right now they are a very surprising and unintuitive one.

**Comment by mhegazy** [comment]
I would not have guessed that NoInfer<T> was the solution :)

**Comment by PyroVortex** [comment]
Unfortunately, using the `U extends T` option has inconsistent behavior for contextual typing, due to the subtly different semantics.
```ts
declare function invoke<T, U extends T, R>(func: (value: T) => R, value: T): R;

declare function test(value: { x: number; }): number;

invoke(test, { x: 1, y: 2 }); // Works
test({ x: 1, y: 2 }); // Compiler error
```

**Comment by RyanCavanaugh** [comment]
Need to collect some use concrete cases before attempting to move forward with this

**Comment by ajafff** [comment]
I also ran into the problem that `U extends T` allows excess properties as @PyroVortex mentions above.

I found a rather simple work around: `{} & T`
The example above would then be:
```ts
declare function invoke<T, R>(func: (value: T) => R, value: {} & T): R;

declare function test(value: { x: number; }): number;

invoke(test, { x: 1, y: 2 }); // Compiler error
test({ x: 1, y: 2 }); // Compiler error
```

**Comment by RyanCavanaugh** [comment]
@ajafff  😲 !

I would expect `{ } & T` to be indistinguishable from `T` - not 100% sure if this is a bug or not but I would not take it as designed behavior.

**Comment by jwbay** [comment]
Ran into this when writing some tests.

```ts
interface User {
    name: string;
    age: number;
}

type ApiCall<T> = (...args: any[]) => Promise<T>;

declare const fetchUser: ApiCall<User>;

declare function mock<T>(call: ApiCall<T>, result: T): void;

//no error, expected missing property (and property completions ideally)
mock(fetchUser, { name: '' });
```

**Comment by RyanCavanaugh** [comment]
Just an update - `T & { } ` creates a "lower-priority" inference site for `T` *by design*. I would move this from the "definitely don't depend on this" column to the "it's probably going to work for the foreseeable future" column.

**Comment by sylvanaar** [comment]
Maybe give it a nice wrapper:

```TypeScript
type InferLast<T> = T & {}
```

Example:

```TypeScript
class Base<TParams> {
	private params: TParams; //cfr https://github.com/Microsoft/TypeScript/issues/1396
}
class Params {
	constructor(public test: string) {
		
	}
}
class Component extends Base<Params> {

}

type InferLast<T> = T & {}

function testInference<TParams>(component: Base<TParams>, params: InferLast<TParams>) {

}

function testInference2<TParams>(component: Base<InferLast<TParams>>, params: TParams) {

}

var component = new Component();

testInference(component, { });   // Error on foo
testInference2(component, { foo:1 });  // Error on component

**Comment by ajafff** [comment]
`T & {}` only works when not using `strictNullChecks`.
I encountered a problem when `T` is `undefined`, because `undefined` is not assignable to `undefined & {}`.

I found that mapped types also defer inference, so i tried `type InferLast<T> = Readonly<T>;`
That does not work for types with index signatures like arrays.

My final solution is a mapped type intersected with the type itself:

```ts
export type NoInfer<T> = T & {[K in keyof T]: T[K]};
```

**Comment by UselessPickles** [comment]
@ajafff FYI - I added your `NoInfer` type to "type-zoo": https://github.com/pelotom/type-zoo/pull/8

**Comment by ajafff** [comment]
With the recent addition of ConditionalTypes also came InferTypes. Currently these can only be used within the condition of the ConditionalType.

Maybe they could be used for regular TypeParameters as well to tell the TypeChecker from which call site it should infer the TypeParameter. That's basically the opposite of the proposal in the opening post.

If you have at least one `infer T` in the signature, the compiler will only use these to infer `T`. The example of the opening post would look like this:

```ts
function doSomething<T>(value: infer T, getDefault: () => T) { }

declare function assertEqual<T>(actual: infer T, expected: T): boolean;
```

Thoughts @RyanCavanaugh?

**Comment by jack-williams** [comment]
@ajafff Would these two cases be equivalent?

```ts
declare function foo<T>(actual: infer T, x: number): T;
declare function foo<T>(actual: T, x: number): T;
```

For consistencies sake it might be nice if a type parameter `T` in a signature _always_ meant the same thing, regardless of context. Maybe do the dual and mark types that should not be inferred, only _checked_.

```ts
function doSomething<T>(value: T, getDefault: () => check T) { }

declare function assertEqual<T>(actual: T, expected: check T): boolean;
```

Parameters marked with `check` never contribute to inference but are only ever checked against. Having a signature with all parameters marked with `check` could either be an error, or be accepted if the user _explicitly_ annotates the call with the instantiation.

```ts
declare function assertEqual<T>(actual: check T, expected: check T): boolean;
assertEqual<number>(4,3); // OK
assertEqual(4,3); // Not OK - a signature with parameters that are only checkable must have an explicit instantiation.
```

**Comment by jwbay** [comment]
I keep running into this. The NoInfer trick with mapped/intersection types works well enough to force correct typechecking, but there are never good completions for it.

```ts
type Component<Props> = {
	props: Props
};

type LowerInference<T> = { // necessary to force an error below
	[P in keyof T]: T[P]
}

declare function doWork<Props>(
	Component: Component<Props>,
	props: LowerInference<Props>,
): any;

declare const comp: Component<{ foo: number }>;

doWork(comp, { // error, property foo is missing 👍
	// but request completions here and get nothing 👎
});
````

**Comment by aleclarson** [comment]
Here's a use case where inference is causing problems: https://goo.gl/g8DrZi

I really like @ajafff's `infer T` syntax.

What's the TS team leaning towards at this time? Would love to see some movement here. :)

**Comment by NN---** [comment]
@ajafff 
> My final solution is a mapped type intersected with the type itself:
> ```ts
> export type NoInfer<T> = T & {[K in keyof T]: T[K]};
> ```

It doesn't solve a simple function.
Where should I put NoInfer to make To type to be non inferred ?


```ts
function convert<To, From>(from: From): To {
    return from as any as To; // doesn't really matter
}

var x = convert(1);  // To = {}
```

**Comment by jcalz** [comment]
How about we use the compiler's deferral of evaluating unresolved conditional types?

```ts
type NoInfer<T> = [T][T extends any ? 0 : never];
```

**Comment by stepanh** [comment]
Linking my SO question here, thanks @jcalz for answering.
https://stackoverflow.com/questions/56687668/a-way-to-disable-type-argument-inference-in-generics

**Comment by AnyhowStep** [comment]
I would like to point out that the three provided workarounds break in the following case,

```ts

class Animal { move }
class Dog extends Animal { woof }
class Cat extends Animal { meow }

type K<T> = T extends any ? {[k in keyof T]:string} : never

export type UnionToIntersection<U> = (
    (
        U extends any ? (k: U) => void : never
    ) extends (
        (k: infer I) => void
    ) ? I : never
);


//https://github.com/microsoft/TypeScript/issues/14829#issuecomment-320833603
type InferLast<T> = T & {}

function doSomethingInferLast<T extends Dog|Cat>(
    getDefault: (x:any) => T,
    value: UnionToIntersection<K<InferLast<T>>>
) { }

//Use a function parameter
//Expected: OK
//Actual: Error
doSomethingInferLast((x) => new Dog(), {woof:"", move:""}); // error on 2nd parameter

//Do not use a function param
//OK
doSomethingInferLast(() => new Dog(), {woof:"", move:""});

///////////

//https://github.com/microsoft/TypeScript/issues/14829#issuecomment-504042546
type NoInfer<T> = [T][T extends any ? 0 : never];

function doSomethingNoInfer<T extends Dog|Cat>(
    getDefault: (x:any) => T,
    value: UnionToIntersection<K<NoInfer<T>>>
) { }

//Use a function parameter
//Expected: OK
//Actual: Error
doSomethingNoInfer((x) => new Dog(), {woof:"", move:""}); // error on 2nd parameter

//Do not use a function param
//OK
doSomethingNoInfer(() => new Dog(), {woof:"", move:""});

/////////////

//https://github.com/microsoft/TypeScript/issues/14829#issuecomment-322267089
export type NoInferAjafff<T> = T & {[K in keyof T]: T[K]};

function doSomethingNoInferAjafff<T extends Dog|Cat>(
    getDefault: (x:any) => T,
    value: UnionToIntersection<K<NoInferAjafff<T>>>
) { }

//Use a function parameter
//Expected: OK
//Actual: Error
doSomethingNoInferAjafff((x) => new Dog(), {woof:"", move:""}); // error on 2nd parameter

//Do not use a function param
//OK
doSomethingNoInferAjafff(() => new Dog(), {woof:"", move:""});

```
[Playground](http://www.typescriptlang.org/play/#code/LAKAxgNghgzjAEBBAdgSwLZQvA3vdA9gG4Cm8AvqJLAgCIEDm8JAHgC4nIAmCKGWueAHcCBAGYUq0OPADCUNs3aceSNJmx50JAkMkhQbAJ4AHMgGkAPABUAfPAC88a0o7cEUZEfgB+XAG0Aa3hUZHhAkiNxZwBdAC4YNgAnUIZyeDj4ZBJSJNBQVhMCJMVjM3gAVTQCZGsCAElkDiSYEjA2VBrLCvsnAApQeCH4AZBh8crXFQ8vXxHAzIqASkd7IgJULgysnJI8seGV1jdVUYnhvoWQ5DE9+HqVhzWNrkHDufrt7NzQJYBufIGEAAemBAAs2GwTDA4qCGKg2GCAK4AIwAdGACOhgehUGAkgQYOI2MDrKYSABlfGoEwk1BwJEkGDAgCMABYABwAJgAnABiekwRmY9DaJoAWgAzFyAAwcyWSgBsMslhnJ9xuewAMrA2DZes54AAyXCUIFiJHIdqdMJcAgUrEkRGpRq3JI6xI2KbueD0BgAH3kbFsZ2GDCdtBIYigSIgbEyfRYcU8Rke9msABo3kMiFhGYtqrUGk09q1rV0rK7tbr9bZbL9BGbQKCKq14FB4BarR0avATFAklBtM1m8CAKIsMztEhcTIAeXMo8Q7SRWEyY6SBP2dodw7BLs17t1fUTaZ2ej9fSWGdwInEcQARA+b4RSI+H+R-vBQcxN8V4L2XLcH2A5Dk6eyAqC9BZAQihIm2HZduWYT9oO6CjguoA7o6zrIAwVZHokJ5ntkF6MFeN44HeYjvi+xAkO+n4AkCoKsWxoKQeCkLQrCwLwoiqIYliOJ4gSRJiCSZJmFSKS0sCgqMsy7LcvyCltFiYpsOKACsMpsnpXLaWyipquUAByxZuvqjjwP41gxHZ3qqCmcwyl8uxJDEzGgEhPa2vaOH7nhFkEV6xzTL6jCBgoIbZvA4ZsJG0axvGIxJimZ6ZnFuYQPmlSFnUjTNGWfmWFYIWHrW9YgCseBNiCwKtmQiGWshIFoeB+yghOU4cLO8CYQ1y5sKuEDrn+24BXuqQVW6J4sCRJBkQwFG3qINFPnRb5Pkx37Ar+W4AWEQFbKhYEjixwLQcgsHwPBzWdq1fntUOGGLiA2HTcFll7MRqznpFK3Xmt96bfg9GMf8nHsTDnEQlCMJwgiyLoiKIn4oSxKkuSMk0nSDJMqynK8gKBMippUpclTioAOxyjyBSTsUpTqrNeyIAAVlAYg89ZTguCaOD+OY1zhJE0T2Zk1jCzE5DeSAvk2vAn1OkFDBs0knPc7zLjhT6frRcGoZDAlSUxnGCbpV4mVZgcOZ5gx+U2oVJYtG0pXlT9mtczzYhVQ2dWcU17aPd2StncOEENT17szvO72gsNo3jVuWFTarM1e1rvvzYty2rVR620eD20fl+P57IdgHARHnWcddt33SHiu9hHb1p7uGffQR2c839TwA5ewOF6Dz4lwxO1Q0CQA)

They break when trying to infer the return type of a function that has at least one parameter.
It's a very specific problem that has been around for a **very** long time.

Specifically,
https://github.com/microsoft/TypeScript/issues/32540

https://github.com/microsoft/TypeScript/issues/29133

Even the `U extends T` workaround breaks. It just thinks that `T` is `Cat|Dog`

**Comment by Validark** [comment]
Here is a repro of an issue I ran into today. I found this when trying to write better types for `Array.reduce`, but that's another story.

```ts
declare function Mutate<T>(callback: (a: T, b: number) => T): T;
Mutate((a, b) => b);
```
**Expected behavior:** assume `T` is `typeof b`, which is `number`

**Actual behavior:** `T` defaults to `unknown`

Another interesting thing about this example is that if you pay attention to the TS tooling, it detects `T` as `number` when you first type `=> b`. However, once it re-evaluates the entire statement it thinks `T` is `unknown`. I don't know the inner workings of TS, but I *guess* that when I typed `=> b` TS was waiting for me to make a better indication of what `T` is, or the Mutate call was already partially evaluated and TS picked up where it left off and this was a side effect of that.

![](https://user-images.githubusercontent.com/15217173/64661729-70943280-d40b-11e9-8a33-611900593002.gif)

**Playground Link:**  https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwFlkMoMQAeAFQD4AKMKCCAIyjAGsAueWqbygDTxm3VMgC2zEDACU8ALzV4lGfwDcAKCIkytXkOZzFwmWqA

**Comment by devinrhode2** [comment]
How are you all creating required generics in your code currently? Are you using @jcalz `NoInfer<T>` trick, or something else?

**Comment by devinrhode2** [comment]
Linking back to a few more SO questions:
First result for "typescript require generic": https://stackoverflow.com/questions/51173191/typescript-require-generic-parameter-to-be-provided
This one is more focused on _classes_: https://stackoverflow.com/questions/57559234/how-to-make-a-generic-of-a-class-is-required

**Comment by kaznovac** [comment]
my question about specifying which generic parameter to infer
https://stackoverflow.com/questions/67108349/is-there-a-way-to-hint-ts-which-occurrence-of-generic-parameter-to-use-for-gener/67110530#67110530

using @jcalz `NoInfer`, as per his suggestion in answer

**Comment by david-shortman** [comment]
> How about we use the compiler's deferral of evaluating unresolved conditional types?
> 
> ```ts
> type NoInfer<T> = [T][T extends any ? 0 : never];
> ```

Idk if this is any better, but a slightly nicer looking version I've come upon looks like this:

```typescript
type NoInfer<T> = T extends infer S ? S : never;
```

**Comment by kugacz** [comment]
> > How about we use the compiler's deferral of evaluating unresolved conditional types?
> > ```ts
> > type NoInfer<T> = [T][T extends any ? 0 : never];
> > ```
> 
> Idk if this is any better, but a slightly nicer looking version I've come upon looks like this:
> 
> ```ts
> type NoInfer<T> = T extends infer S ? S : never;
> ```

Unfortunately, these two types are not equivalent:
```ts
class ExampleClass<T> {}

type NoInferOk<T> = [T][T extends any ? 0 : never];
type NoInferBad<T> = T extends infer S ? S : never;

export class OkClass<T> {
    constructor(private clazz: ExampleClass<T>, private _value: NoInferOk<T>) {
    }

    get value(): T {
        return this._value; // OK. No TS error.
    }    
}

export class BadClass<T> {
    constructor(private clazz: ExampleClass<T>, private _value: NoInferBad<T>) {
    }

    get value(): T {
        return this._value; // TS Error: Type 'NoInferBad<T>' is not assignable to type 'T'. 'T' could be instantiated with an arbitrary type which could be unrelated to 'NoInferBad<T>'.(2322)
    }    
}

```
[Playground link](https://www.typescriptlang.org/play?#code/MYGwhgzhAECiAeYC2AHEBTAwuKAeAKgHzQDeAvgFAUAuAniutAHID2AkgHYBm6ATgPIBrAsQC80ANr4AulOjp41dBwAmMMB1rQA-NAAM0AFzQO6AG59pAbhr1GrTj14AhMCpHRx+eYuVroAJbcfNAAyjphRibmfDYUCigsvNTQoJAwQtjpHiQU0PmpLBwQ1LwArsDUSQAUKLwBZmBKqeAAXq3GCMhoWDgQIgA00HUNTYwA+o0gZejGDsECwkQAlKR5BZTr+QDm6ClTM9XLxt65BefQvHtlvBzQ1AAWARAAdJNg0+hW0AD0P9D8ADSL2YLGg+HCfF4SReW2gZHOFE2CSSKTSUGgrhUWTwRDW52ARRK5UqNRGjWaaXanUQqAwOP6RCG5LG0Henzm7AWWJEqzOGyo512+w+h2O4PxF3yV2oNzuj2ebwOX1+-whcF40N4Jzs0AA5PMnDyiHrAjAOCwUukAtsOGAAEYYe5gugMfX4PUgvUewplEAqaD2xhBEoaagBMYBgDuAUe0A08d49tjvDAvC0rsYUaewAevv9gcYZQ4V3ASgDVX1hr4xsInuqACYAMwNhvLOEIgpIihAA)

**Comment by captain-yossarian** [comment]
> Just an update - `T & { } ` creates a "lower-priority" inference site for `T` _by design_. I would move this from the "definitely don't depend on this" column to the "it's probably going to work for the foreseeable future" column.

@RyanCavanaugh  Is it possible to document this behaviour ?

**Comment by MikeRippon** [comment]
Just linking to this issue I raised about rest parameter inference, as `NoInfer` feels like a pretty clean/readable alternative to requiring more complex behaviour from the compiler: #52227

**Comment by Andarist** [comment]
I've seen cases when `NoInfer` "fails to deliver" its promise though. It would be very cool if this would have builtin support in the compiler :p

**Comment by Andarist** [comment]
@RyanCavanaugh I wonder if, with the recent additions of type parameter modifiers, it wouldn't be possible to use "usage modifiers" over intrinsic types. What I mean is that this could be implemented using a custom syntax/keyword, like this:
```ts
declare function assertEqual<T>(actual: T, expected: noinfer T): boolean;
const g = { x: 3, y: 2 };
assertEqual(g, { x: 3 }); // Error
```

I also wonder what's the current "status" of this proposal. I understand that the TS team is currently not working on this - but what if I would attempt to implement this? I totally understand that any PR is just a PR and nothing is set in stone until merged (and not even that makes anything set in stone 🤣 ) - but I'm hesitant to implement *features* that are likely to be rejected.

**Comment by dinofx** [comment]
I don't see why a new keyword would be needed (`noinfer`).  Maybe the glass is half full:

```ts
declare function invoke<T, R>(func: (value: infer T) => R, value: T): R;
```

`infer` should be unambiguous there, as it isn't part of a conditional extends (related to #52791)

**Comment by Andarist** [comment]
I started working on this [here](https://github.com/microsoft/TypeScript/pull/52968). Feedback is welcome :)

**Comment by Andarist** [comment]
@sandersn has asked me if there are any learnings that I could share based on my PoC implementation of this feature, so here we go:

1. the implementation is pretty straightforward and I didn't encounter many obstacles
2. IMHO an intrinsic type is the best approach for this feature, it's flexible and clearly marks part of the type as "blocked" - with it there are no questions about "operator precedence"-like things
3. It's a powerful tool for declaration authors. Not all libraries would benefit from it but some would. There are a couple of types floating around in the ecosystem that tries to accomplish this. We are using this in XState, I know that RTK is also using this and there are probably more. 
4. I encountered some issues with the custom `NoInfer` (the one proposed by @jcalz). I don't recall what they were now and what was the exact case but in some complex cases, this technique didn't manage to deliver what it should. Having a built-in would set up clear expectations for the behavior and wouldn't rely on implementation details of things like deferral of evaluating unresolved conditional types
5. Further things could be explored later (like `LowInfer`). However, I find the likelihood of that to be quite low. Inferences priorities are also implementation details and a type like this could dangerously leak things to the userland. It's one thing to block the inference completely on given nodes (it doesn't expose much of the internals to the userland) and another thing to give hints about desired inference priorities.

TLDR: TS is insanely expressive already, this would add a new tool to author things that are not always possible today without "hacks" or unintuitive workarounds + the cost of the feature seems to be pretty low to me. What not to like about it? 😉

**Comment by dinofx** [comment]
What happens when `NoInfer` is used in a place where it does nothing?  for example:

```ts
function append<T>(dest: T[], src: NoInfer<T[]>) {
  // TODO
}
```

Where the author intended to do:
```ts
function append<T>(dest: T[], src: NoInfer<T>[]) {
  // TODO
}
```

**Comment by Andarist** [comment]
My PR doesn't assume that this "does nothing". It blocks the whole type from being used as an inference source. You could intentionally wrap a type that references multiple type parameters with a single `NoInfer` type.

**Comment by Andarist** [comment]
I'm not 100% sure about this yet but I also **think** that the builtin `NoInfer` could just perform way better (perf-wise) than the one that we have to use today. In sufficiently complex types TS has to "expand" a lot of types by their constraints and explore all branches to find inference candidates. With `NoInfer` it still has to do that - but it just learns nothing when recursing into types containing it. If a builtin `NoInfer` could wrap "non-leaf" types then TS could bail out early in certain branches - resulting in a better inference performance.

**Comment by jasonkuhrt** [comment]
Just discovered this thread via https://stackoverflow.com/questions/75909231/how-can-this-advanced-function-type-be-achieved/75909852#75909852 where @jcalz enlightened me and pointed out this issue. Thanks for your work @Andarist.

**Comment by jp-diegidio** [comment]
P.S. Sorry, what I have said above, about a single type parameter `T = never` doing the job in case no arguments have type `T`,  is just not true, for example consider this code:

```ts
// for example, just `T` does not appear in the type of the arguments:
function coercion<T = never>(t: unknown): T { return t as T; }

type Num = { x: number };

coercion<Num>([{ x: 1 }]);   // coercion<Num> => OK

coercion([{ x: 1 }])   // coercion<never> => OK, returns `never`!
```

I do have a use case where a plain `T = never` does the trick, but (after looking at it again) it's the specific way I am using `T` to construct the return value that guarantees that I get an error when `T` is `never`...

**Comment by craigphicks** [comment]
@RyanCavanaugh 

> Just an update - `T & { } ` creates a "lower-priority" inference site for `T` _by design_. I would move this from the "definitely don't depend on this" column to the "it's probably going to work for the foreseeable future" column.

Well "noinfer" is here now, but it's also possible to get the error by explicitly constraining the return type of `getDefault` as follows:
```
function doSomething1<T, U>(value:T,getDefault:()=>U&T) { }
// Wanted an error here - getDefault() ought to return same type as 'value'
doSomething1(new Dog(), () => new Animal());
// doSomething1<Dog, unknown>(value: Dog, getDefault: () => Dog): void
```
`T & U` is an entirely logical constraint, explicit and easy to remember, that works just like it should (whereas `T & {}` was a trick).

**Comment by craigphicks** [comment]
> I also ran into the problem that `U extends T` allows excess properties as @PyroVortex mentions above.
> 
> I found a rather simple work around: `{} & T` The example above would then be:
> 
> ```ts
> declare function invoke<T, R>(func: (value: T) => R, value: {} & T): R;
> 
> declare function test(value: { x: number; }): number;
> 
> invoke(test, { x: 1, y: 2 }); // Compiler error
> test({ x: 1, y: 2 }); // Compiler error
> ```

```
declare function invoke<F extends ((value:any) => any)>(func:F, value: Parameters<F>["0"]): ReturnType<F>;
declare function test(value: { x: number; }): number;
invoke(test, { x: 1, y: 2 }); // Compiler Error
//                          ~
test({ x: 1, y: 2 }); // Same Compiler error
//               ~
```

**Comment by olmobrutall** [comment]
I have just updated a big code base to use generic react components thanks to `NoInfer<T>`. So far very happy with the feature. I would just suggest to completely hide `NoInfer<T>` in the tooltip. 

![image](https://github.com/microsoft/TypeScript/assets/720953/8f28b320-2b77-4d8f-9ea1-0788b610e002)

Type declarations are complicated enough already, and `NoInfer<T>` is a hint to the compiler from the library creator but doesn't mean anything for the library consumer.

**Comment by m93a** [comment]
I just wanted to chime in that `NoInfer` saved me from an unintuitive and unexpected `"Type instantiation is excessively deep and possibly infinite."` error when using a function call as a prop value in a complex type. Before using `NoInfer`, TS wanted to infer a generic parameter from the return type of the function, leading to the error. After I marked the function's return value as NoInfer, the error went away!
